src/compute/src/render/reduce.rs

```diff
@@ -800,8 +825,30 @@ where
                     {
                         target.push((row, 1));
                     }
+                } else {
+                    for datum in func
+                        .eval_with_unnest_list::<_, window_agg_helpers::OneByOneAggrImpls>(
+                            iter,
+                            &temp_storage,
+                        )
+                    {
+                        let datum_iter = key.to_datum_iter();
+                        let mut datums_local = datums1.borrow();
+                        datums_local.extend(datum_iter);
+                        let key_len = datums_local.len();
```

Comment:
I think you can avoid decoding the key repeatedly. This assumes that evaluating an mfp only appends datums, but never modifies datums, which I think is true at the moment.

```diff
diff --git a/src/compute/src/render/reduce.rs b/src/compute/src/render/reduce.rs
index c9ceac4470..c74a5b0e90 100644
--- a/src/compute/src/render/reduce.rs
+++ b/src/compute/src/render/reduce.rs
@@ -781,6 +781,7 @@ where
 
         // Allocations for the two closures.
         let mut datums1 = DatumVec::new();
+        let mut datums_key = DatumVec::new();
         let mut datums2 = DatumVec::new();
         let mfp_after1 = mfp_after.clone();
         let mfp_after2 = mfp_after.filter(|mfp| mfp.could_error());
@@ -826,16 +827,17 @@ where
                         target.push((row, 1));
                     }
                 } else {
+                    let mut datums_local = datums_key.borrow();
+                    datums_local.extend(key.to_datum_iter());
+                    let key_len = datums_local.len();
+
                     for datum in func
                         .eval_with_unnest_list::<_, window_agg_helpers::OneByOneAggrImpls>(
                             iter,
                             &temp_storage,
                         )
                     {
-                        let datum_iter = key.to_datum_iter();
-                        let mut datums_local = datums1.borrow();
-                        datums_local.extend(datum_iter);
-                        let key_len = datums_local.len();
+                        datums_local.truncate(key_len);
                         datums_local.push(datum);
                         if let Some(row) = evaluate_mfp_after(
                             &mfp_after1,
```

Reply:
Oh, nice! Thank you!